### PR TITLE
added new methods in controller and used spring jpa

### PR DESCRIPTION
**Pull request by Dima23-ops5**

### CODE REVIEW

File: src/main/java/mate/academy/StringBootIntro/model/Book.java [hunk 1]
```diff
@@ -28,6 +32,8 @@ public class Book {
     private BigDecimal price;
     private String description;
     private String coverImage;
+    @Column(nullable = false)
+    private boolean is_delete = false;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/StringBootIntro/model/Book.java]:
```suggestion
    private boolean isDeleted = false;
```

File: src/main/java/mate/academy/StringBootIntro/model/Book.java [hunk 2]
```diff
@@ -9,11 +8,16 @@
...
 import lombok.Getter;
 import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
+import java.math.BigDecimal;
 
 @Entity
-@Table(name = "books")
+@SQLDelete(sql = "UPDATE books SET is_delete = true WHERE id=?")
+@SQLRestriction(value = "is_delete=false")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/StringBootIntro/model/Book.java]:
```suggestion
@SQLRestriction(value = "is_deleted=false")
```

File: src/main/java/mate/academy/StringBootIntro/model/Book.java [hunk 3]
```diff
@@ -9,11 +8,16 @@
 import jakarta.persistence.Table;
 import lombok.Getter;
 import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
+import java.math.BigDecimal;
 
 @Entity
-@Table(name = "books")
+@SQLDelete(sql = "UPDATE books SET is_delete = true WHERE id=?")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/StringBootIntro/model/Book.java]:
```suggestion
@SQLDelete(sql = "UPDATE books SET is_deleted = true WHERE id=?")
```

File: src/main/java/mate/academy/StringBootIntro/service/BookService.java [hunk 4]
```diff
@@ -5,9 +5,13 @@
...
 
-    public BookDto getBookById(Long id);
+    BookDto getBookById(Long id);
 
-    public BookDto save(CreateBookRequestDto bookDto);
+    BookDto save(CreateBookRequestDto bookDto);
+
+    BookDto updateBookById(Long id, CreateBookRequestDto createBookRequestDto);
+
+    BookDto deleteBookById(Long id);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/StringBootIntro/service/BookService.java]:
```suggestion
    void deleteBookById(Long id);
```

File: src/main/java/mate/academy/StringBootIntro/service/impl/BookServiceImpl.java [hunk 5]
```diff
@@ -26,12 +27,30 @@ public List<BookDto> findAll() {
     @Override
     public BookDto getBookById(Long id) {
         return bookMapper.toDto(bookRepository
-                .getBookById(id).orElseThrow(NoSuchElementException::new));
+                .getById(id));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/StringBootIntro/service/impl/BookServiceImpl.java]: `getById` is deprecated, use `findById`

File: src/main/java/mate/academy/StringBootIntro/service/impl/BookServiceImpl.java [hunk 6]
```diff
@@ -26,12 +27,30 @@ public List<BookDto> findAll() {
     @Override
     public BookDto getBookById(Long id) {
         return bookMapper.toDto(bookRepository
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/StringBootIntro/service/impl/BookServiceImpl.java]:
```suggestion
        return bookRepository.findById(id)
                .map(bookMapper::toDto)
                .orElseThrow(() -> new EntityNotFoundException(....
```

File: src/main/java/mate/academy/StringBootIntro/service/impl/BookServiceImpl.java [hunk 7]
```diff
@@ -26,12 +27,30 @@ public List<BookDto> findAll() {
...
 
     @Override
     public BookDto save(CreateBookRequestDto bookRequetDto) {
         return bookMapper.toDto(bookRepository
                 .save(bookMapper.toModel(bookRequetDto)));
     }
+
+    @Override
+    public BookDto updateBookById(Long id, CreateBookRequestDto createBookRequestDto) {
+        Book book = bookRepository.findById(id)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/StringBootIntro/service/impl/BookServiceImpl.java:42]: Where is a logic of udpating book?

File: src/main/java/mate/academy/StringBootIntro/service/impl/BookServiceImpl.java [hunk 8]
```diff
@@ -26,12 +27,30 @@ public List<BookDto> findAll() {
...
+                .orElseThrow(() -> new NoSuchElementException("Cannot find user with id: "
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/StringBootIntro/service/impl/BookServiceImpl.java]:
```suggestion
                .orElseThrow(() -> new EntityNotFoundException(....
```

Wrong message

File: src/main/java/mate/academy/StringBootIntro/service/impl/BookServiceImpl.java [hunk 9]
```diff
@@ -26,12 +27,30 @@ public List<BookDto> findAll() {
...
+                        + id));
+        return bookMapper.toDto(bookRepository.save(book));
+    }
+
+    @Override
+    public BookDto deleteBookById(Long id) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/StringBootIntro/service/impl/BookServiceImpl.java]:
```suggestion
    public void deleteBookById(Long id) {
        bookRepository.deleteById(id);
    }
```

File: src/main/resources/db/changelog/changes/01-create-book-table.yaml [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+              - column:
+                  name: price
+                  type: bigint
+                  constraints:
+                    nullable: false
+              - column:
+                  name: description
+                  type: varchar(255)
+              - column:
+                  name: coverImage
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/01-create-book-table.yaml]:
```suggestion
                  name: cover_image
```

File: src/main/resources/db/changelog/changes/01-create-book-table.yaml [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+                  type: varchar(255)
+              - column:
+                  name: is_delete
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/01-create-book-table.yaml]:
```suggestion
                  name: is_deleted
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Did you call your endpoints to get result?

File: src/main/java/mate/academy/StringBootIntro/mapper/BookMapper.java [hunk 12]
```diff
@@ -5,10 +5,12 @@
...
 import mate.academy.StringBootIntro.model.Book;
 import org.mapstruct.Mapper;
+import org.mapstruct.MappingTarget;
 
 @Mapper(config = MapperConfig.class)
 public interface BookMapper {
     BookDto toDto(Book book);
 
     Book toModel(CreateBookRequestDto requestDto);
+    void updateBookFromDto(@MappingTarget Book book, CreateBookRequestDto createBookRequestDto);
```
Comment (Vasul-Pankiv) [inline src/main/java/mate/academy/StringBootIntro/mapper/BookMapper.java]: delete it

File: src/main/java/mate/academy/StringBootIntro/service/impl/BookServiceImpl.java [hunk 13]
```diff
@@ -25,13 +27,30 @@ public List<BookDto> findAll() {
...
     public BookDto save(CreateBookRequestDto bookRequetDto) {
         return bookMapper.toDto(bookRepository
                 .save(bookMapper.toModel(bookRequetDto)));
     }
+
+    @Override
+    public BookDto updateBookById(Long id, CreateBookRequestDto createBookRequestDto) {
+        Book book = bookRepository.findById(id)
+                .orElseThrow(EntityNotFoundException::new);
+        bookMapper.updateBookFromDto(book, createBookRequestDto);
```
Comment (Vasul-Pankiv) [inline src/main/java/mate/academy/StringBootIntro/service/impl/BookServiceImpl.java]: use repository to update book

File: src/main/resources/db/changelog/changes/01-create-book-table.yaml [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+                  name: description
+                  type: varchar(255)
+              - column:
+                  name: cover_image
+                  type: varchar(255)
+              - column:
+                  name: is_deleted
+                  type: tinyint
+                  constraints:
+                    nullable: false
```
Comment (Vasul-Pankiv) [inline src/main/resources/db/changelog/changes/01-create-book-table.yaml]: add empty line to the end of file

File: src/main/resources/liqubase.properties [hunk 15]
```diff
@@ -0,0 +1,4 @@
+url=jdbc:mysql://localhost:3306/BookStore
+username=root
+password=bookPass
+changelog-file=classpath:db/changelog/db.changelog-master.yaml
```
Comment (Vasul-Pankiv) [inline src/main/resources/liqubase.properties]: same

**Review by Vasul-Pankiv — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/StringBootIntro/service/impl/BookServiceImpl.java [hunk 16]
```diff
@@ -25,13 +27,34 @@ public List<BookDto> findAll() {
...
+    @Override
+    public BookDto updateBookById(Long id, CreateBookRequestDto createBookRequestDto) {
+        Book book = bookRepository.findById(id)
+                .orElseThrow(EntityNotFoundException::new);
+       book.setPrice(createBookRequestDto.getPrice());
+       book.setCoverImage(createBookRequestDto.getCoverImage());
+       book.setIsbn(createBookRequestDto.getIsbn());
+       book.setTitle(createBookRequestDto.getTitle());
+       book.setAuthor(createBookRequestDto.getAuthor());
+        return bookMapper.toDto(book);
```
Comment (sarakhmen) [inline src/main/java/mate/academy/StringBootIntro/service/impl/BookServiceImpl.java]: why don't you save your book after updating its fields?

File: src/main/java/mate/academy/StringBootIntro/model/Book.java [hunk 17]
```diff
@@ -9,11 +8,16 @@
...
+import java.math.BigDecimal;
 
-@Entity
-@Table(name = "books")
 @Getter
 @Setter
+@Entity
+@Table(name = "books")
+@SQLDelete(sql = "UPDATE books SET isDeleted = true WHERE id=?")
+@SQLRestriction(value = "isDeleted=false")
```
Comment (sarakhmen) [inline src/main/java/mate/academy/StringBootIntro/model/Book.java]: your column name is 'is_deleted'

File: src/main/java/mate/academy/StringBootIntro/mapper/BookMapper.java [hunk 18]
```diff
@@ -5,10 +5,11 @@
...
 import org.mapstruct.Mapper;
+import org.mapstruct.MappingTarget;
 
 @Mapper(config = MapperConfig.class)
 public interface BookMapper {
     BookDto toDto(Book book);
 
     Book toModel(CreateBookRequestDto requestDto);
-}
+   }
```
Comment (sarakhmen) [inline src/main/java/mate/academy/StringBootIntro/mapper/BookMapper.java]: style issue. Please fix such issue in all files

**Review by sarakhmen — CHANGES_REQUESTED** [review]
let's improve your solution :)

File: src/main/java/mate/academy/StringBootIntro/controller/BookController.java [hunk 19]
```diff
@@ -29,8 +26,20 @@ public BookDto getBookById(@PathVariable Long id) {
...
 
     @PostMapping
-    public BookDto createBook(@RequestBody CreateBookRequestDto bookDto) {
+    public BookDto createBook(@RequestBody @Valid CreateBookRequestDto bookDto) {
         return bookService.save(bookDto);
     }
 
+    @PutMapping("/{id}")
+    public BookDto updateBookByID(@PathVariable Long id,
+                                  @RequestBody CreateBookRequestDto createBookRequestDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/StringBootIntro/controller/BookController.java:35]:
```suggestion
                                  @RequestBody @Valid CreateBookRequestDto createBookRequestDto) {
```

File: src/main/java/mate/academy/StringBootIntro/controller/BookController.java [hunk 20]
```diff
@@ -29,8 +26,20 @@ public BookDto getBookById(@PathVariable Long id) {
...
+        return bookService.updateBookById(id, createBookRequestDto);
+    }
+
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    @DeleteMapping("/{id}")
+    public void deleteBookById(@PathVariable Long id) {
+        bookService.deleteBookById(id);
+    }
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/StringBootIntro/controller/BookController.java:44]:
```suggestion
```

File: src/main/java/mate/academy/StringBootIntro/model/Book.java [hunk 21]
```diff
@@ -27,18 +31,8 @@ public class Book {
     @Column(nullable = false)
     private BigDecimal price;
     private String description;
+    @Column(name = "cover_image")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/StringBootIntro/model/Book.java:34]:
```suggestion
```
Hibernate automatically generate column in snake_case style

File: src/main/java/mate/academy/StringBootIntro/service/impl/BookServiceImpl.java [hunk 22]
```diff
@@ -25,13 +27,34 @@ public List<BookDto> findAll() {
 
     @Override
     public BookDto getBookById(Long id) {
-        return bookMapper.toDto(bookRepository
-                .getBookById(id).orElseThrow(NoSuchElementException::new));
+        return bookMapper.toDto(bookRepository.findById(id)
+                .orElseThrow(EntityNotFoundException::new));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/StringBootIntro/service/impl/BookServiceImpl.java:31]: Add informative message

File: src/main/java/mate/academy/StringBootIntro/service/impl/BookServiceImpl.java [hunk 23]
```diff
@@ -25,13 +27,34 @@ public List<BookDto> findAll() {
...
     @Override
     public BookDto save(CreateBookRequestDto bookRequetDto) {
         return bookMapper.toDto(bookRepository
                 .save(bookMapper.toModel(bookRequetDto)));
     }
+
+    @Override
+    public BookDto updateBookById(Long id, CreateBookRequestDto createBookRequestDto) {
+        Book book = bookRepository.findById(id)
+                .orElseThrow(EntityNotFoundException::new);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/StringBootIntro/service/impl/BookServiceImpl.java:43]: the same

File: src/main/java/mate/academy/StringBootIntro/service/impl/BookServiceImpl.java [hunk 24]
```diff
@@ -25,13 +27,34 @@ public List<BookDto> findAll() {
...
+       book.setPrice(createBookRequestDto.getPrice());
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/StringBootIntro/service/impl/BookServiceImpl.java:44]: use Mapper to update book

File: src/main/java/mate/academy/StringBootIntro/service/impl/BookServiceImpl.java [hunk 25]
```diff
@@ -25,13 +27,34 @@ public List<BookDto> findAll() {
...
+       book.setCoverImage(createBookRequestDto.getCoverImage());
+       book.setIsbn(createBookRequestDto.getIsbn());
+       book.setTitle(createBookRequestDto.getTitle());
+       book.setAuthor(createBookRequestDto.getAuthor());
+        return bookMapper.toDto(bookRepository.save(book));
+    }
+
+    @Override
+    public void deleteBookById(Long id) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/StringBootIntro/service/impl/BookServiceImpl.java:53]:
```suggestion
    public void deleteBookById(Long id) {
        bookRepository.deleteById(id);
    }
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Let's fix some comments

**Pull request closed without merging**